File: JavaDemo/src/work-spring/acmicpc/src/main/java/com/grepp/boot/controller/UserController.java
package com.grepp.boot.controller;

import com.grepp.boot.model.dto.UserDTO;
import com.grepp.boot.model.service.UserService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpSession;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;

@Controller
@RequestMapping("/users")
public class UserController {
    @Autowired
    private UserService userService;

    @GetMapping("/login")
    public String loginForm() {
        return "loginForm";
    }

    @PostMapping("/login")
    public ModelAndView login(@RequestParam String userId, @RequestParam String userPw, HttpSession session, HttpServletRequest request) {
        ModelAndView mav = new ModelAndView("alert");

        UserDTO user = userService.getUser(userId);
        if (user != null && user.getUserPw().equals(userPw)) {
            session.setAttribute("loginUser", user);
            mav.addObject("msg", "로그인 성공 !!");
            mav.addObject("path", request.getContextPath() + "/main");
        } else {
            mav.addObject("msg", "로그인 실패 8ㅅ8");
            mav.addObject("path", request.getContextPath() + "/users/login");
        }

        return mav;
    }

    @GetMapping("/logout")
    public ModelAndView logout(HttpSession session, HttpServletRequest request) {
        session.invalidate();

        ModelAndView mav = new ModelAndView("alert");
        mav.addObject("msg", "로그아웃 되었습니다.");
        mav.addObject("path", request.getContextPath() + "/users/login");

        return mav;
    }
}

File: JavaDemo/src/work-kotlin/day1021_todo/src/main/kotlin/com/grepp/todo/controller/UserController.kt
package com.grepp.todo.controller

import com.grepp.todo.model.dto.UserDTO
import com.grepp.todo.model.entity.UserEntity
import com.grepp.todo.model.service.UserService
import com.grepp.todo.util.MyJwtTokenProvider
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/auth")
class UserController(val service: UserService) {
    // autowired는 객체 생성 이후 주입이기 때문에 반드시 lateinit var 해줘야 함.
    @Autowired
    lateinit var myTokenProvider: MyJwtTokenProvider

    // 회원가입
    @PostMapping("/signup")
    fun register(@RequestBody userDTO: UserDTO): ResponseEntity<UserEntity> {
        val result = service.join(userDTO.toEntity())
        println("join result : ${result}")
        result.role = "ADMIN"   // 지금 권한이나 소셜 정보 하나도 구현 안했기 때문에..
        result.authProvider = "kakao"
        return ResponseEntity.ok().body(result)
    }

    // 로그인
    @PostMapping("/login")
    fun login(@RequestBody userDTO: UserDTO): ResponseEntity<Any> {
        val user = service.login(userDTO.username, userDTO.password)

        println("login result : ${user}")
        if (user != null) {     // 로그인 성공했어? 그럼 토큰줘야지~
            val token = myTokenProvider.createMyToken(user, 1000*60*2)
            userDTO.token = token
            return ResponseEntity.ok().body(user)
        }

        return ResponseEntity.badRequest().body("login failed!!")
    }
}